prefetcher/EIP/EIP.cc

```diff
@@ -0,0 +1,871 @@
+} l1i_hist_entry;
+
+l1i_hist_entry l1i_hist_table[NUM_CPUS][L1I_HIST_TABLE_ENTRIES];
+uint64_t l1i_hist_table_head[NUM_CPUS];      // log_2 (L1I_HIST_TABLE_ENTRIES)
+uint64_t l1i_hist_table_head_time[NUM_CPUS]; // 64 bits
+
+void l1i_init_hist_table()
+{
+  l1i_hist_table_head[l1i_cpu_id] = 0;
+  l1i_hist_table_head_time[l1i_cpu_id] = current_core_cycle[l1i_cpu_id];
+  for (uint32_t i = 0; i < L1I_HIST_TABLE_ENTRIES; i++) {
+    l1i_hist_table[l1i_cpu_id][i].tag = 0;
+    l1i_hist_table[l1i_cpu_id][i].time_diff = 0;
+    l1i_hist_table[l1i_cpu_id][i].bb_size = 0;
+  }
+}
+
+uint64_t l1i_find_hist_entry(uint64_t line_addr)
+{
+  uint64_t tag = line_addr & L1I_HIST_TAG_MASK;
+  for (uint32_t count = 0, i = (l1i_hist_table_head[l1i_cpu_id] + L1I_HIST_TABLE_MASK) % L1I_HIST_TABLE_ENTRIES; count < L1I_HIST_TABLE_ENTRIES;
+       count++, i = (i + L1I_HIST_TABLE_MASK) % L1I_HIST_TABLE_ENTRIES) {
+    if (l1i_hist_table[l1i_cpu_id][i].tag == tag)
+      return i;
+  }
+  return L1I_HIST_TABLE_ENTRIES;
+}
+
+// It can have duplicated entries if the line was evicted in between
+void l1i_add_hist_table(uint64_t line_addr)
+{
+  // Insert empty addresses in hist not to have timediff overflows
+  while (current_core_cycle[l1i_cpu_id] - l1i_hist_table_head_time[l1i_cpu_id] >= L1I_TIME_DIFF_OVERFLOW) {
+    l1i_hist_table[l1i_cpu_id][l1i_hist_table_head[l1i_cpu_id]].tag = 0;
+    l1i_hist_table[l1i_cpu_id][l1i_hist_table_head[l1i_cpu_id]].time_diff = L1I_TIME_DIFF_MASK;
+    l1i_hist_table[l1i_cpu_id][l1i_hist_table_head[l1i_cpu_id]].bb_size = 0;
+    l1i_hist_table_head[l1i_cpu_id] = (l1i_hist_table_head[l1i_cpu_id] + 1) % L1I_HIST_TABLE_ENTRIES;
+    l1i_hist_table_head_time[l1i_cpu_id] += L1I_TIME_DIFF_MASK;
+  }
+
+  // Allocate a new entry (evict old one if necessary)
+  l1i_hist_table[l1i_cpu_id][l1i_hist_table_head[l1i_cpu_id]].tag = line_addr & L1I_HIST_TAG_MASK;
+  l1i_hist_table[l1i_cpu_id][l1i_hist_table_head[l1i_cpu_id]].time_diff =
+      (current_core_cycle[l1i_cpu_id] - l1i_hist_table_head_time[l1i_cpu_id]) & L1I_TIME_DIFF_MASK;
+  l1i_hist_table[l1i_cpu_id][l1i_hist_table_head[l1i_cpu_id]].bb_size = 0;
+  l1i_hist_table_head[l1i_cpu_id] = (l1i_hist_table_head[l1i_cpu_id] + 1) % L1I_HIST_TABLE_ENTRIES;
+  l1i_hist_table_head_time[l1i_cpu_id] = current_core_cycle[l1i_cpu_id];
+}
+
+void l1i_add_bb_size_hist_table(uint64_t line_addr, uint32_t bb_size)
+{
+  uint64_t index = l1i_find_hist_entry(line_addr);
+  l1i_hist_table[l1i_cpu_id][index].bb_size = bb_size & L1I_MERGE_BBSIZE_MAX_VALUE;
+}
+
+// compute delta
+uint32_t l1i_find_bb_merge_hist_table(uint64_t line_addr)
+{
+  uint64_t tag = line_addr & L1I_HIST_TAG_MASK;
+  for (uint32_t count = 0, i = (l1i_hist_table_head[l1i_cpu_id] + L1I_HIST_TABLE_MASK) % L1I_HIST_TABLE_ENTRIES; count < L1I_HIST_TABLE_ENTRIES;
+       count++, i = (i + L1I_HIST_TABLE_MASK) % L1I_HIST_TABLE_ENTRIES) {
+    if (count >= L1I_BB_MERGE_ENTRIES) {
+      return 0;
+    }
```

Comment:
What is this doing here? Isn't this always the terminating condition for the loop, since `L1I_BB_MERGE_ENTRIES < L1I_HIST_TABLE_ENTRIES`?